Third lab

Adi6e opened the pull request:
Добавил ЛР3. В 3.3 я дополнительно построил приближающий полином третьей степени, чтобы показать, что коэффициенты полинома считаются правильно, так как сумма квадратичных ошибок для полиномов 1 и 2 степени одинаковые с точностью до 1e-10.
P.S. Если что, ldarlene - это я, если вдруг возникнут вопросы, почему новые ветки создает не Adi6e, а ldarlene.

Adi6e committed:
first commit

Adi6e committed:
Merge pull request #1 from Adi6e/main

first commit

Adi6e committed:
Revert "first commit"

Adi6e committed:
Merge pull request #2 from pivovarov-mai/revert-1-main

Revert "first commit"

Adi6e committed:
first commit

Lab1

Adi6e committed:
lab2 first commit

Adi6e committed:
ЛР3 пункты 3.1 и 3.2

Adi6e committed:
Delete Lab3 from pull request for Lab2

Commit:
added lab3

Commit:
added lab3

Adi6e committed:
Delete Lab3 directory

Adi6e committed:
small changes in 3.3

Изменил вывод программы 3.3 .

pivovarov-mai commented on Lab3/3.1/newton_and_lagrange.py:
```diff
@@ -0,0 +1,83 @@
+        print("Error")
+
+
+def divided_difference(X):
+    if len(X) == 1:
+        return f(X[0])
+    return (divided_difference(X[:-1]) - divided_difference(X[1:])) / (X[0] - X[-1])
```
Что это такое и будет ли это верно для массивов длины большей 3?

Adi6e replied:
Это рекурсивный подсчёт, работать будет для длин больше 3, потому что я считаю в том числе f(x1, x2, x3, x4) и результат верный

pivovarov-mai replied:
Рекурсию я заметил ), а подсчет чего?

pivovarov-mai commented on Lab3/3.2/cubic_spline.py:
```diff
@@ -0,0 +1,83 @@
+def find_range(X, x_point):
+    ind = 0
+    for i in range(len(X)):
+        if X[i] >= x_point:
```
вещественные числа на ЭВМ лучше не сравнивать нестрогим неравенством.

Adi6e replied:
Мне переделать эту функцию?

pivovarov-mai replied:
оставляйте, на будущее имейте ввиду

pivovarov-mai commented on Lab3/3.4/derivative.py:
```diff
@@ -0,0 +1,51 @@
+        print("There is no such point!")
+        return
+    elif i == 0:
+        return 0
```
А если левой точки нет, правильно ли, что производная равна нулю?

Adi6e committed:
changes in 3.2 and 3.4

Переделал функции подсчета левосторонней и правосторонней производной в 3.4, исправил функцию нахождения точки в диапазоне в 3.2
